src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -376,41 +382,38 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
+                ImmutableArray<AnalyzerWithState> analyzersWithState,
                 AnalysisKind kind,
                 TextSpan? span,
                 ArrayBuilder<DiagnosticData> builder,
                 bool incrementalAnalysis,
                 CancellationToken cancellationToken)
             {
                 Debug.Assert(!incrementalAnalysis || kind == AnalysisKind.Semantic);
-                Debug.Assert(!incrementalAnalysis || stateSetsAndExistingData.All(stateSetAndData => stateSetAndData.stateSet.Analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis()));
+                Debug.Assert(!incrementalAnalysis || analyzersWithState.All(analyzerWithState => analyzerWithState.Analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis()));
 
-                using var _ = ArrayBuilder<StateSet>.GetInstance(stateSetsAndExistingData.Length, out var stateSetBuilder);
-                foreach (var (stateSet, existingData) in stateSetsAndExistingData)
+                using var _ = ArrayBuilder<AnalyzerWithState>.GetInstance(analyzersWithState.Length, out var filteredAnalyzersWithStateBuilder);
+                foreach (var analyzerWithState in analyzersWithState)
                 {
-                    var analyzer = stateSet.Analyzer;
-                    if (_priorityProvider.MatchesPriority(analyzer))
-                    {
-                        // Check if this is an expensive analyzer that needs to be de-prioritized to a lower priority bucket.
-                        // If so, we skip this analyzer from execution in the current priority bucket.
-                        // We will subsequently execute this analyzer in the lower priority bucket.
-                        Contract.ThrowIfNull(existingData);
-                        if (await TryDeprioritizeAnalyzerAsync(analyzer, existingData.Value).ConfigureAwait(false))
-                        {
-                            continue;
-                        }
+                    Debug.Assert(_priorityProvider.MatchesPriority(analyzerWithState.Analyzer));
 
-                        stateSetBuilder.Add(stateSet);
+                    // Check if this is an expensive analyzer that needs to be de-prioritized to a lower priority bucket.
+                    // If so, we skip this analyzer from execution in the current priority bucket.
+                    // We will subsequently execute this analyzer in the lower priority bucket.
+                    if (await TryDeprioritizeAnalyzerAsync(analyzerWithState.Analyzer, analyzerWithState.ExistingData).ConfigureAwait(false))
+                    {
+                        continue;
                     }
+
+                    filteredAnalyzersWithStateBuilder.Add(analyzerWithState);
                 }
 
-                var stateSets = stateSetBuilder.ToImmutable();
+                analyzersWithState = filteredAnalyzersWithStateBuilder.ToImmutable();
```

Comment:
if it's empty, you'll get the singleton `ImmutableArray<T>.Empty` value

Reply:
I was looking at this through code inspection, so maybe I'm looking at the wrong code? 

ArrayBuilder<T>
        public ImmutableArray<T> ToImmutable()
        {
            return _builder.ToImmutable();
        }

ImmutableArray<T>.Builder
            public ImmutableArray<T> ToImmutable()
            {
                return new ImmutableArray<T>(this.ToArray());
            }

Reply:
I agree that the extension should probably handle the empty case and avoid an allocation. However, for the case here we don't expect all the analyzers to get deprioritized, so for almost all cases this builder will be non-empty.

Reply:
Let me add the reverse optimization - if the filtered set is same as the original one, we avoid an allocation.

Reply:
This new code confuses me. If nothing is deprioritized, it uses the analyzersWithState passed into the method to determine whether to return?

I feel like it would be simpler to just get rid of the anyDeprioritized flag, and do something like this after the loop:

if (filteredAnalyzersWithStateBuilder.Count == 0)
    return;

analyzersWithState = filteredAnalyzersWithStateBuilder.ToImmutable();

Reply:
Done, thanks.